libvirt: Convert igniton-config file to config drive image on s390x/s390

gyohuangxin opened the pull request:
On the s390x/s390 platform, the igniton-config file will be converted to config
 drive image, so that the terraform can use it to inject metadata into
 coreos image.

 Related: https://github.com/dmacvicar/terraform-provider-libvirt/pull/666#issue-340880431

 [Distribution via Config Drive]: https://coreos.com/os/docs/latest/config-drive.html

patrickdillon commented:
/label platform/libvirt

jichenjc commented on pkg/asset/store/store.go:
```diff
@@ -357,3 +372,27 @@ func (s *storeImpl) purge(excluded []asset.WritableAsset) error {
+	if err != nil {
+		return err
+	}
+	cmd = exec.Command("mkisofs", "-R", "-V", "config-2", "-o", "configdrive.iso", ".")
```
are you sure it's mkisofs? I think it's genisoimage?

gyohuangxin replied:
I think they makes the same effect.

tuan-hoang1 replied:
Maybe related:
https://github.com/weldr/lorax/blob/master/share/templates.d/99-generic/s390.tmpl#L79

jichenjc commented on pkg/asset/store/store.go:
```diff
@@ -83,6 +85,19 @@ func (s *storeImpl) Fetch(a asset.Asset, preserved ...asset.WritableAsset) error
+				return err
+			}
+		}
+	}
```
this is much like a hacking, do we get other ways? just curious...

gyohuangxin replied:
Yes, I'm trying other ways.

jaypoulz replied:
I know that we are targeting this change specific to s390x, but would it be more acceptable to target this change for all remote libvirt installations where fw_cfg is not available?
I'm not sure how you make a test for that.

jaypoulz replied:
Prashanth wanted me to note that we should be careful to ensure this doesn't run when the install type is OpenStack, since it already has an config drive, but this looks specific to libvirt.

gyohuangxin replied:
@jaypoulz Thanks, I think it's a good thought. However, before we use config drive, a configdrive iso should be created, so that terraform can use. The iso file should be created by installer.

> Prashanth wanted me to note that we should be careful to ensure this doesn't run when the install type is OpenStack, since it already has an config drive, but this looks specific to libvirt.

Yes, it effects only when install type is libvirt.

crawford commented on pkg/asset/store/store.go:
```diff
@@ -83,6 +85,19 @@ func (s *storeImpl) Fetch(a asset.Asset, preserved ...asset.WritableAsset) error
+	// On the s390x/s390 platform, the igniton-config file will be converted to config
+	// drive image, so that the terraform can use it to inject metadata.
+	if runtime.GOARCH == "s390x" || runtime.GOARCH == "s390" && a.Config.Libvirt != nil {
+		if reflect.TypeOf(a).String() == "*bootstrap.Bootstrap" ||
```
Why include this in the store and not the Ignition assets?

gyohuangxin replied:
@crawford Thanks, I agree. I will change in next update.

jaypoulz replied:
It was brought up today in the CI/CD meeting that the node that does in the install is running on an x86_64 cluster, whereas the target node is s390x. I think you will need a check here to pull in the remote arch.

crawford replied:
I'm not sure if it's possible to determine the remote arch (cc @abhinavdahiya). We might have to define a new platform or maybe just add a field to the existing libvirt platform options (i.e. `UseConfigDrive` under the libvirt `Platform` struct).

gyohuangxin replied:
Remote arch can be determined by libvirt api which used by libvirt-go. However, installer should add the libvirt-go dependence, I don't think it's a good method.
Could we set an env variable (i.e. `TARGET`) to tell installer the target node's arch?

[terraform-provider-libvirt](https://github.com/dmacvicar/terraform-provider-libvirt) and [cluster-api-provider-libvirt](https://github.com/openshift/cluster-api-provider-libvirt) has libvirt-go, so they can expose apis to installer. 

In addition, we are trying if all these steps can be moved to terraform-provider-libvirt and cluster-api-provider-libvirt.

crawford commented on pkg/asset/store/store.go:
```diff
@@ -357,3 +372,27 @@ func (s *storeImpl) purge(excluded []asset.WritableAsset) error {
+	if err := os.MkdirAll("openstack/latest", 0755); err != nil {
+		return err
+	}
+	cmd := exec.Command("cp", file, "openstack/latest/user_data")
```
Don't shell out to utilities when you can copy the file using Go directly.

crawford commented on pkg/asset/store/store.go:
```diff
@@ -357,3 +372,27 @@ func (s *storeImpl) purge(excluded []asset.WritableAsset) error {
+	if err != nil {
+		return err
+	}
+	cmd = exec.Command("mkisofs", "-R", "-V", "config-2", "-o", "configdrive.iso", ".")
```
I don't think you need to bother with ISOs. Ignition is just looking for `/dev/disk/by-label/config-2`, so any filesystem with that label will work.

gyohuangxin replied:
@crawford Do you mean that the suffix ".iso" is not needed? I'll have a try.

jichenjc replied:
openstack itself only support iso or vfat.. but seems the ignition doesn't have this restriction
https://github.com/coreos/ignition/blob/master/internal/providers/openstack/openstack.go#L117
so we should be able to avoid iso from code here

gyohuangxin replied:
> openstack itself only support iso or vfat.. but seems the ignition doesn't have this restriction
> https://github.com/coreos/ignition/blob/master/internal/providers/openstack/openstack.go#L117
> so we should be able to avoid iso from code here

@jichenjc  Yes, I agree.

gyohuangxin replied:
> I don't think you need to bother with ISOs. Ignition is just looking for `/dev/disk/by-label/config-2`, so any filesystem with that label will work.

@crawford The ignition reads the filesystem's [CONFIG-2/config-2 label](https://github.com/coreos/ignition/blob/a924dd7d11685a30bf79f20ce9fc3ca93860f032/internal/providers/openstack/openstack.go#L75), how can we make a label to a file without ISOs. 

I've found a method like below and I'll have a try. https://docs.openstack.org/ironic/latest/install/configdrive.html#accessing-the-configuration-drive-data
```
mkdir -p /openstack/latest/user_data
mount /dev/disk/by-label/config-2 /openstack/latest/user_data
```

crawford replied:
udev populates those symlinks based on the filesystem label. All you need is a filesystem which supports labelling (e.g. ext4, xfs, vfat).

> how can we make a label to a file without ISOs

You can't - at least, not in the initramfs. You'll need to rely on the mechanism mentioned above to get the config to Ignition.

I originally made the suggestion in the hopes that it might simplify the implementation and maybe drop some dependencies (I was hoping someone had put together a Golang implementation for FAT that we could just use to create the filesystem ourselves). If this alternate route ends up being more complicated, then I would abandon it. We'll need to make sure that the `mkisofs` utility is available in the test environment and the MAO container image, if we continue down this route.

jichenjc replied:
right, avoid the utils will be a good thing, do we have a fixed format for our disks ? otherwise seems the ext2/3/4/ using e2progs and xfs using xfs_admin.. so maybe we need consider the logic here as well.. @gyohuangxin

gyohuangxin commented:
@crawford @jaypoulz I've updated the code, moved the method to Ignition assets and set a env variable `USE_CONFIG_DRIVE`. When `USE_CONFIG_DRIVE` is True, installer will create config-drive iso in a tmp directory.

After that, installer create the main.tf and set the content as below to use the config-drive iso. 

```
resource "libvirt_ignition" "bootstrap" {
  name    = "${var.cluster_id}-bootstrap.ign"
  content = "/tmp/bootstrap.iso"
  pool    = var.pool
}
```
At last terraform will mount the iso at images to inject igniton file, related to https://github.com/dmacvicar/terraform-provider-libvirt/pull/666.

jichenjc commented on cmd/openshift-install/create.go:
```diff
@@ -169,6 +170,13 @@ func runTargetCmd(targets ...asset.WritableAsset) func(cmd *cobra.Command, args
 				return err
 			}
+
+			if os.Getenv("USE_CONFIG_DRIVE") == "True" {
```
can we rename CONFIG_DRIVE to IGNITION_FORCE_USE_CONFIG_DRIVE or something similar to make it more specific?

gyohuangxin replied:
Sure~

jichenjc commented on pkg/asset/asset.go:
```diff
@@ -101,3 +103,53 @@ func isDirEmpty(name string) (bool, error) {
+		}
+
+		// copy the config file as /openstack/latest/user_data
+		if err2 := os.Chdir(directory); err2 != nil {
```
why we have this err2?

gyohuangxin replied:
Sorry, my mistake.

jichenjc commented on pkg/asset/asset.go:
```diff
@@ -101,3 +103,53 @@ func isDirEmpty(name string) (bool, error) {
+
+		logrus.Debugf("Creating config-drive image for %s", filename)
+		// mkisofs will create an ISO 9660 filesystem labeled 'config-2'
+		cmd := exec.Command("mkisofs", "-R", "-V", "config-2", "-o", "/tmp/"+isoname, "/tmp/new-drive/")
```
I don't know whether we need document somewhere, but mkisofs seems not in my machine by default..

wking commented on pkg/asset/targets/targets.go:
```diff
@@ -66,6 +66,9 @@ var (
 		&password.KubeadminPassword{},
+		&machine.Master{},
+		&machine.Worker{},
+		&bootstrap.Bootstrap{},
```
What are these additions for?

gyohuangxin replied:
Because when we use command `bin/openshift-install create cluster`, the master.ign, bootstrap.ign and worker.ign will not be generated in form of file,  these additions in targets.go tell installer to generate them, so that iso file can be created.

jichenjc replied:
I used this and only see following , but master ign should be very big?  XXXX below is the sshkey I am having..

{"ignition":{"config":{"append":[{"source":"https://api-int.test1.aa.testing:22623/config/master","verification":{}}]},"security":{"tls":{"certificateAuthorities":[{"source":"data:text/plain;charset=utf-8;base64,XXXXXXXXXXXXXXXXXXXXX","verification":{}}]}},"timeouts":{},"version":"2.2.0"},"networkd":{},"passwd":{},"storage":{},"systemd":{}}

gyohuangxin replied:
@jichenjc 
I think what you said may be bootstrap.ign instead of master.ign. 
I run installer in the original way, below is master ignition's details in terraform.tfvars.json.
terraform.tfvars.json contains all info installer tells terraform, and master ignition is small as yours.

```
  "ignition_master": "{\"ignition\":{\"config\":{\"append\":[{\"source\":\"https://api-int.test1.tt.testing:22623/config/master\",\"verification\":{}}]},\"security\":{\"tls\":{\"certificateAuthorities\":[{\"source\":\"data:text/plain;charset=utf-8;base64,XXXXXXXXXX",\"verification\":{}}]}},\"timeouts\":{},\"version\":\"2.2.0\"},\"networkd\":{},\"passwd\":{},\"storage\":{},\"systemd\":{}}"
```

gyohuangxin commented:
@crawford Hi, could we have a further discuss about this pr?

So far, we use a environment variable `USE_IGNITION_CONFIG_DRIVE=True` to tell installer to use this code. Then installer generates ignition files and create config-drive iso labeled `config-2`. We use the `mkisofs` utility which you concerned about, and I confirmed that there is only `genisoimage` in container such as cluster-api-operator, refer to https://github.com/openshift/cluster-api-provider-libvirt/blob/master/Dockerfile. So we will change `mkisofs` to `genisoimage`, they make the same effect.

After that, terraform will mount iso images to vms in form of  cdrom, related: https://github.com/dmacvicar/terraform-provider-libvirt/pull/666#issue-340880431

crawford commented:
I opened up an alternative PR to the terraform-provider-libvirt changes: https://github.com/dmacvicar/terraform-provider-libvirt/pull/678. If we go with this approach, we won't need these changes to the installer. I know that the alternative PR still uses, `mkisofs`, but I did that because that dependency already existed in terraform-provider-libvirt. We can add that utility to the necessary container images.

gyohuangxin commented:
> I opened up an alternative PR to the terraform-provider-libvirt changes: [dmacvicar/terraform-provider-libvirt#678](https://github.com/dmacvicar/terraform-provider-libvirt/pull/678). If we go with this approach, we won't need these changes to the installer. I know that the alternative PR still uses, `mkisofs`, but I did that because that dependency already existed in terraform-provider-libvirt. We can add that utility to the necessary container images.

Thanks a lot, we will try it in our env.

crawford commented:
Closing in favor of https://github.com/dmacvicar/terraform-provider-libvirt/pull/678.
<!--
/close
-->
